core/states/approaching_gate.py

```diff
@@ -0,0 +1,195 @@
+from core.vision.ar_tag_detector import is_gate
+import core
+import interfaces
+import algorithms
+from core.states import RoverState
+import time
+import math
+
+
+class ApproachingGate(RoverState):
+    """
+    Within approaching marker, the rover explicitly follows the spotted marker until it reaches an acceptable distance from the rover, or loses sight of it.
+    """
+
+    def start(self):
+        # Schedule AR Tag detection
+        self.num_detection_attempts = 0
+        self.gate_detection_attempts = 0
+
+    def exit(self):
+        # Cancel all state specific coroutines
+        pass
+
+    def on_event(self, event) -> RoverState:
+        """
+        Defines all transitions between states based on events
+        """
+        state: RoverState = None
+
+        if event == core.AutonomyEvents.REACHED_MARKER:
+            state = core.states.Idle()
+
+        elif event == core.AutonomyEvents.START:
+            state = self
+
+        elif event == core.AutonomyEvents.MARKER_UNSEEN:
+            state = core.states.SearchPattern()
+
+        elif event == core.AutonomyEvents.ABORT:
+            state = core.states.Idle()
+
+        else:
+            self.logger.error(f"Unexpected event {event} for state {self}")
+            state = core.states.Idle()
+
+        # Call exit() if we are not staying the same state
+        if state != self:
+            self.exit()
+
+        # Return the state appropriate for the event
+        return state
+
+    async def run(self) -> RoverState:
+
+        # Call AR Tag tracking code to find position and size of AR Tag
+        if core.vision.ar_tag_detector.is_gate():
+            # Grab the AR tags
+            tags = core.vision.ar_tag_detector.get_tags()
+            gps_data = core.waypoint_handler.get_waypoint()
+            orig_goal, orig_start, leg_type = gps_data.data()
+
+            # If we've seen at least 5 frames of 2 tags, assume it's a gate
+            print(leg_type, self.gate_detection_attempts)
+            if len(tags) == 2 and self.gate_detection_attempts >= 5 and leg_type == "GATE":
+                self.logger.info("Gate detected, beginning navigation")
+                # compute the angle across from the gate
+                # depending where the rover is facing, this is computed differently
+                if tags[0].angle < 0 and tags[1].angle < 0:  # both tags on the right
+                    larger = min(tags[0].angle, tags[1].angle)
+                    smaller = max(tags[0].angle, tags[1].angle)
+                    combinedAngle = abs(larger) - abs(smaller)
+                elif tags[0].angle >= 0 and tags[1].angle >= 0:  # both tags on the left
+                    larger = max(tags[0].angle, tags[1].angle)
+                    smaller = min(tags[0].angle, tags[1].angle)
+                    combinedAngle = larger - smaller
+                else:  # one tag on left, one on right
+                    combinedAngle = abs(tags[0].angle) + abs(tags[1].angle)
```

Comment:
Maybe consider pre-computing and naming the result of these conditionals (which are used multiple times). For example, calling this heading = "north_south" or something like that may make the code a lot more legible.

Reply:
Moving this to its own issue, a separate project to clean up the math